controllers/providers/metrics/singleton.go

```diff
@@ -0,0 +1,51 @@
+/*
+Copyright 2021 The k8gb Contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+Generated by GoLic, for more details see: https://github.com/AbsaOSS/golic
+*/
+package metrics
+
+import (
+	"sync"
+
+	"github.com/AbsaOSS/k8gb/controllers/depresolver"
+)
+
+const (
+	// DefaultMetricsNamespace provides the default namespace used, when PrometheusMetrics was not initialised
+	DefaultMetricsNamespace = "k8gb_default"
+)
+
+var (
+	o       sync.Once
+	metrics = nop()
```

Comment:
Good question, thx:

first of all, the logger / metrics recorder should have a default state even if the `metrics.Init(config)` function is not called yet. This doesn't matter during runtime, because `Init(config)` is registered soon after [loading configuration](https://github.com/k8gb-io/k8gb/blob/prometheus_no3_metrics_global_object/main.go#L102).

The problem is when we run a random test - for example a provider test where Init(config) is not explicitly called. Metrics will be called on nil object and an error will occur. Similarly, logging, which is used across the application, passes through tests without having to explicitly initialize it before each test. And that's because Nop() inits logger silently.
 
This way, the metric always has a default (empty but not nil) behavior ([uses this prefix](https://github.com/k8gb-io/k8gb/blob/3d19cfac835cdeb1bd0244630a74ea968d726849/controllers/providers/metrics/singleton.go#L28)), and the behavior of the metric can only be changed once by calling Init(config).

Reply:
By calling nop in the var or init section it is guaranteed that the object will be initialized right after the package is loaded, so I don't have to worry about whether the metrics is nil or not

Reply:
Very cool, thanks for the explanation!